view/frontend/web/js/view/payment/method-renderer/omise-offsite-truemoney-method.js

```diff
@@ -0,0 +1,126 @@
+define(
```

Comment:
Noticed that a lot of these `method-renderer` JS files are mostly the same. There'd probably be scope to make this more efficient? Feels very wrong to be repeating so much code

Reply:
i don't see solution for that actually. magento 2 is designed this way (it is used knockout js to push rendering from backend side to frontend side).
we should keep separated renderers because some of the parts are different. if you would see that all of them are objects it will be a mess to add in those renderers some extra conditions depends on what payment methods they concern.

Reply:
Well, the code seems to just be calling a `define` function, and passing an array of strings and a function. The things that are passed seem almost the same in each `method-renderer` JS file. Couldn't we pull that array out to somewhere else, along with a function that generates the required function for passing in? So, in the `method-renderer` JS files, we would have something like:

```
define(DEPENDENCY_ARRAY, generateMethodRendererFunction('truemoney'))
```

My quick example is based upon looking at the Alipay and TrueMoney method renderer JS files - which are almost identical

Reply:
that is reinventing how Magento is recommending to do it @jonrandy

Reply:
by any chance, did you find any example that supports your way of doing that changes on internet? magento official developer website?

Reply:
It's common programming sense. I'm not changing anything about how Magento does things - simply removing unnecessary code duplication. We have very large blocks of code here that are all but identical except for a few strings. We would be calling the same Magento functions using the same arguments. Nothing changes except for more making it more maintainable, succinct code

Reply:
ok, thank you for your comment @jonrandy. any refactoring that concerns any other payment method will be done in separate PR.

Reply:
BTW @jonrandy your bit emotional response and a quite big comment about the _common programming sense_ was not answering my question.
You answered:
_`It's common programming sense. I'm not changing anything about how Magento does things - simply removing unnecessary code duplication. We have very large blocks of code here that are all but identical except for a few strings - that's a really bad code smell. We would be calling the same Magento functions using the same arguments. Nothing changes except for more making it more maintainable, succinct code`_

I asked if you found anything how to do this stuff in Magento with `knockout JS`
in existing Magento architecture. it is easy to throw things.. and ask people .. `yeah, you check it!` I will just give you some ideas like:
`define(DEPENDENCY_ARRAY, generateMethodRendererFunction('truemoney')).`
because of _code smells_ for you.

Reply:
Your question about Magento/Knockout examples did not make sense in relation to my comments, as my comments did not relate to Magento or Knockout. I am merely pointing out that in this part of the code we have very large sections of code that are being repeated. This is a situation that programming as a discipline in general seeks to remove.

I am not for one second advocating changing the way Knockout or Magento are being used, and I am not reinventing anything. I am merely suggesting that we improve our code to remove this unnecessary duplication. I will put together a related example to explain further

Reply:
Upon closer inspection of the JS files there, it appears they are using Require JS. It would be relatively simple to split the 'method renderers' that we have into a class hierarchy with parent classes implementing the common code, and moving specifics into child classes. My proposed solution would work, but it circumvents the RequireJS mechanisms already available to us. Doing this spiltting would greatly improve the maintainability of the code, and reduce amount of code and effort required to add new payment methods on Magento 2. This is a worthwhile exercise and we should definitely do it.

Reference: https://devdocs.magento.com/guides/v2.3/javascript-dev-guide/javascript/js_overview.html

Reply:
thank you @jonrandy I will have a look at it.